### ISSUE TITLE
libjulia functions

### ISSUE THREAD

**Issue by tshort** [MikeInnes/Charlotte.jl#4]
Where in the pipeline would one swap out libjulia functions for the wasm/js equivalent? 

Is that just adding to the `wasmfuncs` dict? 

That dict doesn't have the argument types, so it seems like something different might be needed to swap out functions that have multiple (dispatch) methods.

**Comment by MikeInnes** [comment]
As in, `jl_` ccalls in the IR? I've basically only come across intrinsics so far so that's what `wasmfuncs` is targeted at. `wasmcall` is slightly more primitive than `wasmfunc` so you can do more with that, or in the worst case you can just add a branch [here](https://github.com/MikeInnes/Charlotte.jl/blob/fe52726bb2e754a35f3f122dac3cda2292de7edd/src/wasm/compile.jl#L208) with some arbitrary transformation.

**Comment by tshort** [comment]
Yes, that's for when you hit `ccall`s and in a few cases `llvmcall`s, but we might want to intercept some calls earlier in the calling sequence. You hit this for a lot of things like Arrays. Swapping out some things like error checking also helps get code to work in some cases.

**Comment by tshort** [comment]
Also, great stuff! I like your overall plan. Many of the approaches I've tried run into roadblocks that required me to push to lower and lower levels of processing `code_typed` code anyway, so I'm all for trying this. This seems like a good approach for generating small wasm. If you have a to-do list that you need help with, I'd be happy to pitch in.

**Comment by MikeInnes** [comment]
That's a bit harder. Usually these functions get inlined, at which point it's not very easy to intercept them (we've hit this a lot with CUDAnative as well). I think at some point the compiler is going to have to provide some kind of api for replacing functions wholesale. Until then, just replacing e.g. `throw` with a trap should get us by.

Glad you like it! I've opened issues here and on WebAssembly.jl for all the immediate things I can think of. Any suggestions on top of that are more than welcome – if you play around with it you'll no doubt run into things that could be improved.

**Comment by tshort** [comment]
Calling JavaScript functions may be another use of `ccall`. Or, maybe we need `jscall`.
